lib/stream_data.ex

```diff
@@ -505,18 +505,31 @@ defmodule StreamData do
   end
 
   defp integer_lazy_tree(int, range) do
-    next_division = fn
-      0 -> nil
-      n -> {n, div(n, 2)}
-    end
+    lazy_tree(int, &integer_lazy_tree(int, range, _current = int, &1, &2))
+  end
+
+  defp integer_lazy_tree(_int, _range, _current, {:halt, acc}, _fun) do
+    {:halted, acc}
+  end
 
-    children =
-      int
-      |> Stream.unfold(next_division)
-      |> Stream.drop_while(&((int - &1) not in range))
-      |> Stream.map(&integer_lazy_tree(int - &1, range))
+  defp integer_lazy_tree(int, range, current, {:suspend, acc}, fun) do
+    {:suspended, acc, &integer_lazy_tree(int, range, current, &1, fun)}
+  end
+
+  defp integer_lazy_tree(int, range, current, {:cont, acc}, fun) do
+    to_emit = int - current
+
+    cond do
+      to_emit == int ->
+        {:done, acc}
 
-    lazy_tree(int, children)
+      to_emit in range ->
```

Comment:
If we're going for max speed here, it might make sense to deconstruct the rage and pass `start, end` in the loop - this should save a protocol dispatch.

Reply:
@michalmuskala this PR is directed at speeding up the creation of the lazy tree returned by `integer/0,1`, not the "actuation" of said lazy tree. We only "realize" this tree when we shrink and go down the tree, otherwise we only use the root; however, creating a few streams turns out to be way more expensive than returning a function (`&integer_lazy_tree(...)`) that does the streaming. Hope I made this a bit clearer :)